File: cpp/brokerage.cpp
#include <map>
#include <iostream>
#include <string>

class Brokerage {
	public:
		float cash = 10000.0;
		std::map<std::string, float> holdings;
		void buy(std::string ticker, std::map<std::string, float> ticker_prices, std::string date, float dollar_amount);
		void sell(std::string ticker, std::map<std::string, float> ticker_prices, std::string date, float dollar_amount);
		float Brokerage::account_total(std::map<std::string, std::map<std::string, float>> ticker_prices, std::string date);
};

void Brokerage::buy(std::string ticker, std::map<std::string, float> ticker_prices, std::string date, float dollar_amount) {
	if (dollar_amount > Brokerage::cash)
		throw std::invalid_argument("Not enough cash to buy");

	if (ticker_prices.count(date) == 0)
		throw std::invalid_argument("Invalid date");

	if (Brokerage::holdings.count(ticker) == 0)
		Brokerage::holdings[ticker] = 0;

	float date_price = ticker_prices[date];
	float amount_purchased = dollar_amount / date_price;

	Brokerage::cash -= dollar_amount;
	Brokerage::holdings[ticker] += amount_purchased;

}

void Brokerage::sell(std::string ticker, std::map<std::string, float> ticker_prices, std::string date, float dollar_amount) {
	if (Brokerage::holdings.count(ticker) == 0)
		throw std::invalid_argument("Stock not even owned nerd");
	
	if (ticker_prices.count(date) == 0)
		throw std::invalid_argument("Invalid date");


	float date_price = ticker_prices[date];
	float stock_amount_selling = dollar_amount / date_price;

	if (stock_amount_selling > Brokerage::holdings[ticker])
		throw std::invalid_argument("Trying to sell more than what's in holdings");

	Brokerage::cash += dollar_amount;
	Brokerage::holdings[ticker] -= stock_amount_selling;

}

float Brokerage::account_total(std::map<std::string, std::map<std::string, float>> all_ticker_prices, std::string date){
	float stock_value = 0;
	for (const auto& pair : Brokerage::holdings) {
    	const std::string& ticker = pair.first;
    	float amount = pair.second;

		std::map<std::string, float> ticker_prices = all_ticker_prices[ticker];
		float current_stock_price = ticker_prices[date];
		stock_value += current_stock_price * amount;
	}

	return Brokerage::cash + stock_value;
}


int main() {
	printf("test\n");

	return 0;
}

File: python/brokerage.py
import pandas as pd
import datetime


'''IDEAS
        -Make verbose option that prints out whenever a purchase is made, or better yet, prints to a file
        -Make functions that show stuff like return or growth over time or something idk
        -Make __str__ that prints the holdings dictionary pretty
        
        -Eventually will need to track how many trades each ticker has and how long we held each ticker so that we can incorporate management/holding and trading fees
'''

class Brokerage():
    def __init__(self, cash = 10000):
        
        #Establishes the dictionary that keeps track of everything
        #Cash is in dollars, every security is in amount of stocks owned
        self.cash = cash
        self.holdings = {}


    def buy(self, ticker, ticker_dataframe, date, dollar_amount = 0, percentage = 0):
        #HANDLE THESE SOMEWHERE
        #Makes sure we have enough cash to buy
        if dollar_amount > self.cash:
            raise ValueError
        
        #Makes sure date is valid
        if date not in ticker_dataframe['Date'].tolist():
            raise ValueError    #should we raise a different error to distinguish between the two?
        
        #Makes sure we only chose one purchase method
        if dollar_amount > 0 and percentage > 0:
            raise ValueError
        
        #This allows us below to keep adding on investments into the same thing, otherwise, we would need to reset everytime at 0.
        try:
            self.holdings[ticker]
        except:
            self.holdings[ticker] = 0
        
        #Finds the price at said date and total cash_used
        date_price = self.price_from_date(date, ticker_dataframe)
        cash_used = self.cash * percentage + dollar_amount  #since one must be zero, this should work for both cases
        
        #calculates how much we are purchasing
        amount_purchased = cash_used / date_price

        #resets our cash amount and sets our new holding amount purchased, makes sure to not reset 
        self.cash = self.cash - cash_used
        self.holdings[ticker] = self.holdings[ticker] + amount_purchased



    def sell(self, ticker, ticker_dataframe, date, dollar_amount = 0, percentage = 0):
        #HANDLE THESE SOMEWHERE
        #Makes sure we even hold the stock
        try:
            self.holdings[ticker]
        except:
             ValueError
        
        #Makes sure date is valid
        if date not in ticker_dataframe['Date'].tolist():
            raise ValueError    #should we raise a different error to distinguish between the two?
        
        #Makes sure we only chose one sell method
        if dollar_amount > 0 and percentage > 0:
            raise ValueError
        
        #Finds the price at said date
        date_price = self.price_from_date(date, ticker_dataframe)
    
        #finds how many stocks we're selling
        stock_amount_selling = self.holdings[ticker] * percentage + dollar_amount/date_price  #since one must be zero, this should work for both cases
        cash_selling = stock_amount_selling * date_price

        #Makes sure we aren't trying to sell more then we have
        if self.holdings[ticker] < stock_amount_selling:
            raise ValueError

        #updates our cash amount and resets our holding amount purchased, makes sure to not reset 
        self.cash = self.cash + cash_selling
        self.holdings[ticker] = self.holdings[ticker] - stock_amount_selling



    def price_from_date(self, date, dataframe):
        i = dataframe.index[dataframe["Date"] == date].tolist()
        return dataframe[i]["Close"]
